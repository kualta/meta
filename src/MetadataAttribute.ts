import { z } from 'zod';

import { NonEmptyStringSchema } from './primitives.js';

/**
 * The type of a metadata attribute.
 */
export enum MetadataAttributeType {
  BOOLEAN = 'Boolean',
  DATE = 'Date',
  NUMBER = 'Number',
  STRING = 'String',
  JSON = 'JSON',
}

export type BooleanAttribute = {
  /**
   * A JS boolean value serialized as string. It's consumer responsibility to parse it.
   */
  value: 'true' | 'false';
  /**
   * Union discriminant.
   */
  type: MetadataAttributeType.BOOLEAN;
  /**
   * The attribute's unique identifier.
   */
  key: string;
};
/**
 * @internal
 */
export const BooleanAttributeSchema = z.object({
  type: z.literal(MetadataAttributeType.BOOLEAN),
  key: NonEmptyStringSchema.describe("The attribute's unique identifier."),
  value: z
    .enum(['true', 'false'])
    .describe("A JS boolean value serialized as string. It's consumer responsibility to parse it."),
});

export type DateAttribute = {
  /**
   * A valid ISO 8601 date string.  It's consumer responsibility to parse it.
   */
  value: string;
  /**
   * Union discriminant.
   */
  type: MetadataAttributeType.DATE;
  /**
   * The attribute's unique identifier.
   */
  key: string;
};
/**
 * @internal
 */
export const DateAttributeSchema = z.object({
  type: z.literal(MetadataAttributeType.DATE),
  key: NonEmptyStringSchema.describe("The attribute's unique identifier."),
  value: z
    .string()
    .datetime()
    .describe("A valid ISO 8601 date string.  It's consumer responsibility to parse it."),
});

export type NumberAttribute = {
  /**
   * A valid JS number serialized as string. It's consumer responsibility to parse it.
   *
   * @example
   * ```ts
   * '42'
   *
   * '42n'
   *
   * '42.42'
   * ```
   */
  value: string;
  /**
   * Union discriminant.
   */
  type: MetadataAttributeType.NUMBER;
  /**
   * The attribute's unique identifier.
   */
  key: string;
};
/**
 * @internal
 */
export const NumberAttributeSchema = z.object({
  type: z.literal(MetadataAttributeType.NUMBER),
  key: NonEmptyStringSchema.describe("The attribute's unique identifier."), // TODO generalize and share
  value: NonEmptyStringSchema.describe(
    "A valid JS number serialized as string. It's consumer responsibility to parse it.",
  ),
});

export type StringAttribute = {
  /**
   * Any string value.
   */
  value: string;
  /**
   * Union discriminant.
   */
  type: MetadataAttributeType.STRING;
  /**
   * The attribute's unique identifier.
   */
  key: string;
};
/**
 * @internal
 */
export const StringAttributeSchema = z.object({
  type: z.literal(MetadataAttributeType.STRING),
  key: NonEmptyStringSchema.describe("The attribute's unique identifier."),
  value: NonEmptyStringSchema.describe('Any string value.'),
});

export type JSONAttribute = {
  /**
   * A JSON string. It's consumer responsibility to validate and parse it.
   */
  value: string;
  /**
   * Union discriminant.
   */
  type: MetadataAttributeType.JSON;
  /**
   * Union discriminant.
   */
  /**
   * The attribute's unique identifier.
   */
  key: string;
};
/**
 * @internal
 */
export const JSONAttributeSchema = z.object({
  type: z.literal(MetadataAttributeType.JSON),
  key: NonEmptyStringSchema.describe("The attribute's unique identifier."),
  value: NonEmptyStringSchema.describe(
    "A JSON string. It's consumer responsibility to validate and parse it.",
  ),
});

/**
 * A Lens metadata attribute.
 */
export type MetadataAttribute =
  | BooleanAttribute
  | DateAttribute
  | NumberAttribute
  | StringAttribute
  | JSONAttribute;

export const MetadataAttributeSchema = z.discriminatedUnion('type', [
  BooleanAttributeSchema,
  DateAttributeSchema,
  NumberAttributeSchema,
  StringAttributeSchema,
  JSONAttributeSchema,
]);
