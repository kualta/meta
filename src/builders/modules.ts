import type { MetadataAttribute } from '../MetadataAttribute';
import { type ModuleMetadata, ModuleMetadataSchema, ModuleSchemaId } from '../module';
import type { Prettify } from '../utils';
import { evaluate } from './ValidationError';
import type { RecursiveUnbrand } from './utils';

/**
 * @private
 * @privateRemarks MUST stay very @private to produce usable docs
 */
type ModuleDetails = Prettify<RecursiveUnbrand<Omit<ModuleMetadata, '$schema'>>>;
/**
 * All {@link ModuleMetadata} fields without `$schema` a with:
 * - `attributes` defaults to an empty array
 */
export type ModuleOptions = ModuleDetails & {
  /**
   * An arbitrary bag of attributes you wish to add to the metadata.
   */
  attributes?: MetadataAttribute[];
};

/**
 * Creates a valid ModuleMetadata.
 *
 * @category Compose
 * @param input - Use your IDE suggestions for an enhanced development experience
 *
 * @example
 * ```ts
 * const metadata = module({
 *   name: 'My Open Action',
 *   title: 'This is my Open Action',
 *   description: 'Get ready for the future of social interaction!',
 *   authors: ['awesome-dev@lens.xyz'],
 *   initializeCalldataABI: JSON.stringify([
 *     {
 *       type: 'address',
 *       name: 'address',
 *     },
 *     {
 *       type: 'uint256',
 *       name: 'price',
 *     },
 *   ]),
 *   processCalldataABI: JSON.stringify([
 *     {
 *       type: 'address',
 *       name: 'collector',
 *     },
 *   ]),
 * });
 * ```
 */
export function module({ attributes = [], ...others }: ModuleOptions): ModuleMetadata {
  return evaluate(
    ModuleMetadataSchema.safeParse({
      $schema: ModuleSchemaId.LATEST,
      attributes,
      ...others,
    }),
  );
}
