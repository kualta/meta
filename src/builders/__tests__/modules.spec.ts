import { describe, expect, it } from 'vitest';

import { MetadataAttributeType } from '../../MetadataAttribute.js';
import { module } from '../modules.js';

describe(`Given the "${module.name}" metadata builder`, () => {
  describe('when using it', () => {
    it('should return a valid OpenActionMetadata', () => {
      const metadata = module({
        name: 'Open Action Name',
        title: 'Open Action Title',
        authors: ['me@lens.xyz'],
        attributes: [
          {
            key: 'twitter',
            type: MetadataAttributeType.STRING,
            value: 'https://twitter.com/johndoe',
          },
          {
            key: 'dob',
            type: MetadataAttributeType.STRING,
            value: '1990-01-01T00:00:00Z',
          },
        ],
        description: `
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        - Donec a diam lectus.
        `,
        initializeCalldataABI: JSON.stringify([
          {
            type: 'address',
            name: 'address',
            indexed: true,
          },
          {
            type: 'uint256',
            name: 'price',
            indexed: true,
          },
        ]),
        processCalldataABI: JSON.stringify([
          {
            type: 'address',
            name: 'collector',
            indexed: true,
          },
        ]),
      });

      expect(metadata).toMatchSnapshot();
    });
  });
});
