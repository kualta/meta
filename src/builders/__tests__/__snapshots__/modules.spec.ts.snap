// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the "module" metadata builder > when using it > should return a valid OpenActionMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/modules/1.0.0.json",
  "attributes": [
    {
      "key": "twitter",
      "type": "String",
      "value": "https://twitter.com/johndoe",
    },
    {
      "key": "dob",
      "type": "String",
      "value": "1990-01-01T00:00:00Z",
    },
  ],
  "authors": [
    "me@lens.xyz",
  ],
  "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        - Donec a diam lectus.",
  "initializeCalldataABI": "[{"type":"address","name":"address","indexed":true},{"type":"uint256","name":"price","indexed":true}]",
  "name": "Open Action Name",
  "processCalldataABI": "[{"type":"address","name":"collector","indexed":true}]",
  "title": "Open Action Title",
}
`;
